### DISCUSSION TITLE
Determining appropriate pnpm version to use from lock file

### DISCUSSION THREAD

**Discussion by privatenumber** [pnpm/pnpm#3341] [Q&A]
I use `npx pnpm i --frozen-lockfile` in my CIs to install dependencies, but with the new pnpm v6 release that drops Node.js v10 support, installations have been breaking in projects where I test in a Node.js v10 environment.

While I am happy that Node.js v10 support is being dropped, I can imagine this being a problem again with future Node.js version deprecations.

Of course I can work around this by manually fixing the version of pnpm to install, but I was wondering if there's a "smarter" way to automatically determine which version of pnpm to use.

I checked `pnpm-lock.yaml` for any hints of which version of pnpm was used to create it, but the only hint I seem to get is `lockfileVersion`, which doesn't seem to directly correspond with the pnpm version.

Is there a way to infer or determine the version of pnpm used to create the lock file?

**Comment by zkochan** [comment]
Currently, in the docs, we recommend explicitly install [a major pnpm version on CI](https://pnpm.io/continuous-integration), like: 

```
curl -f https://get.pnpm.io/v6.js | node - add --global pnpm@6
```

For dev environment, I think we will recommend to use [volta](https://volta.sh/) which will automatically pick the right pnpm version (but it does not support pnpm yet).

As for automatically installing the right pnpm version globally... I guess we could implement something like it but I am not sure how needed that is. I think it is pretty straightforward to specify which pnpm version needs to be installed.

**Reply by privatenumber** [reply]
If not currently possible, no worries. I agree it's not a major issue either.

I was also wondering if I can add some sort of pnpm version detection logic in [`ci`](https://github.com/privatenumber/ci), but I can just hardcode which version of Node.js is supported with which package manager versions.

Thank you @zkochan for the fast response and your hard work on pnpm -- it's everything I wish npm was!

**Reply by zkochan** [reply]
Well, maybe we could publish some distribution tags, so you could do something like `install -g pnpm@node12`

**Reply by privatenumber** [reply]
I test with different Node.js versions using GitHub Action's matrix feature [like this](https://github.com/privatenumber/esbuild-loader/blob/d9bc73b28337d7ec4a78b36318b2e363c96f6975/.github/workflows/test.yml#L23), so I don't think there's a simple way to say install from `pnpm@node12` without adding extra logic to make it complex.

I ended up adding [this logic in `npx ci`](https://github.com/privatenumber/ci/blob/15b526d340b003c3cef38b166acdbde916e13e1d/ci.js#L28-L29), so my issue is resolved.

Thank you for your consideration!
